### PR TITLE
feat: option to query metrics only on polling interval

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubalik@gmail.com>

Adding an option in the trigger spec, that enables caching metric values during polling interval. Then each request coming to KEDA Metrics Server from k8s server (HPA controller), reads the value from this cache and not querying the external service directly. This reduces load on the external services and user have the ability to select this option for an individual trigger in ScaledObject.
- this feature is **EXPERIMENTAL** and I'd like to gather feedback, before making it a stable one
- this feature is available only for ScaledObjects
- I named this property ~`queryMetricsOnPollingInterval`~, though I am not happy with the name and I am eager to hear suggestions -> `useCachedMetrics`

See example:
```yaml
  - type: kafka
    useCachedMetrics: true     # <-- NEW OPTION, default value: false
    metadata:
      bootstrapServers: kafka.svc:9092
      consumerGroup: my-group
      topic: test-topic
      lagThreshold: '5'
```

Code changes:
- I merged `scaler.IsActive()` and `scaler.GetMetrics()` calls to single method `scaler.GetMetricsAndActivity()` -> this way we don't ask for the same information twice. I was able to reduce this calls for majority of scalers, there are some leftovers to be refactored, as the logic was more complex (so in `scaler.GetMetricsAndActivity()` we call `scaler.IsActive()` and `scaler.GetMetrics()` methods instead of single call) - TODO tracking issue for leftovers
- refactored scalehandler and scalerscache a little bit

### Outstanding issues:
- [x] How to name this property?
- [x] e2e test
- [x] more unit tests
- [x] print warning message if this message is being used in ScaledJobs - it doesn't work there
- [x] pring error if this is used by scaler where it doesn't make sense - cpu, memory, cron Where else?

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs/pull/1003
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes https://github.com/kedacore/keda/issues/3921

Relates to https://github.com/kedacore/keda/issues/2282

### CODE REVIEW

**Comment by tomkerkhove** [comment]
What about "useCachedMetrics" or "useCachedMetricsDuringPollingInterval"?

**Comment by JorTurFer** [comment]
`useCachedMetrics` looks good

File: controllers/keda/scaledobject_controller_test.go [hunk 1]
```diff
@@ -52,671 +46,6 @@ func (g GinkgoTestReporter) Fatalf(format string, args ...interface{}) {
...
-			err = k8sClient.Create(context.Background(), so)
-			Ω(err).ToNot(HaveOccurred())
-
-			Eventually(func() metav1.ConditionStatus {
-				err = k8sClient.Get(context.Background(), types.NamespacedName{Name: soName, Namespace: "default"}, so)
-				Ω(err).ToNot(HaveOccurred())
-				return so.Status.Conditions.GetReadyCondition().Status
-			}, 20*time.Second).Should(Equal(metav1.ConditionFalse))
-		})
-	})
```
Comment (JorTurFer) [inline controllers/keda/scaledobject_controller_test.go:719]: why did you remove these tests?

Comment (zroubalik) [reply controllers/keda/scaledobject_controller_test.go:719]: just temporary stuff, for easier testing, will put that back

**Comment by v-shenoy** [comment]
> @JorTurFer @v-shenoy ready for a review

Will take a look by EOD.

File: tests/internals/cache_metrics/cache_metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	data.MonitoredDeploymentReplicas = 4
+	KubectlApplyWithTemplate(t, data, "monitoredDeploymentTemplate", monitoredDeploymentTemplate)
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, 4, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	t.Log("waiting 30s to allow propagation of changes")
+
+	// Let's sleep at least 30s
+	// the usual setting for `horizontal-pod-autoscaler-sync-period` is 15s)
+	time.Sleep(30 * time.Second)
```
Comment (JorTurFer) [inline tests/internals/cache_metrics/cache_metrics_test.go:218]: do we need this here? the `WaitForDeploymentReplicaReadyCount` already waits some time. We could just extend the timeout from 1 minute to 2. WDYT?

File: tests/internals/cache_metrics/cache_metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
```
Comment (JorTurFer) [inline tests/internals/cache_metrics/cache_metrics_test.go:177]: I think we can use the helper function `AssertReplicaCountNotChangeDuringTimePeriod` to ensure that replica count doesn't change. Internally, the helper waits some time and evaluates the deployment replica count. WDYT?

**Review by JorTurFer — COMMENTED** [review]
Awesome work! I have left some comments inline

File: pkg/scalers/azure_eventhub_scaler.go [hunk 4]
```diff
@@ -431,3 +431,20 @@ func (s *azureEventHubScaler) Close(ctx context.Context) error {
...
+	}
+
+	isActive, err := s.IsActive(ctx)
+	if err != nil {
+		s.logger.Error(err, "error getting activity status")
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+
+	return metrics, isActive, nil
+}
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply pkg/scalers/azure_eventhub_scaler.go:450]: That's the plan, to open a issue once this get merged. It is more complicated and I think it can wait, or can be tackled separately

File: pkg/scalers/external_scaler.go [hunk 5]
```diff
@@ -217,6 +217,23 @@ func (s *externalScaler) GetMetrics(ctx context.Context, metricName string) ([]e
...
+
+	isActive, err := s.IsActive(ctx)
+	if err != nil {
+		s.logger.Error(err, "error getting activity status")
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+
+	return metrics, isActive, nil
+}
+
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply pkg/scalers/external_scaler.go:236]: That's the plan, to open a issue once this get merged. It is more complicated and I think it can wait, or can be tackled separately

File: pkg/scalers/predictkube_scaler.go [hunk 6]
```diff
@@ -239,6 +239,23 @@ func (s *PredictKubeScaler) GetMetrics(ctx context.Context, metricName string) (
...
+
+	isActive, err := s.IsActive(ctx)
+	if err != nil {
+		s.logger.Error(err, "error getting activity status")
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+
+	return metrics, isActive, nil
+}
+
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply pkg/scalers/predictkube_scaler.go:258]: That's the plan, to open a issue once this get merged. It is more complicated and I think it can wait, or can be tackled separately

File: Makefile [hunk 7]
```diff
@@ -53,7 +53,7 @@ GO_LDFLAGS="-X=github.com/kedacore/keda/v2/version.GitCommit=$(GIT_COMMIT) -X=gi
 COSIGN_FLAGS ?= -a GIT_HASH=${GIT_COMMIT} -a GIT_VERSION=${VERSION} -a BUILD_DATE=${DATE}
 
 # ENVTEST_K8S_VERSION refers to the version of kubebuilder assets to be downloaded by envtest binary.
-ENVTEST_K8S_VERSION = 1.24
+ENVTEST_K8S_VERSION = 1.25
```
[the remark this answers is not in the document]
Comment (JorTurFer) [reply Makefile:56]: I check that usually , so I can do it once it's stable :)  Let's keep 1.24.6 for the moment

File: pkg/scaling/scale_handler.go [hunk 8]
```diff
@@ -270,15 +281,17 @@ func (h *scaleHandler) GetScalersCache(ctx context.Context, scalableObject inter
...
+	withTriggers, err := kedav1alpha1.AsDuckWithTriggers(scalableObject)
 	if err != nil {
 		return err
 	}
 
 	key := withTriggers.GenerateIdentifier()
 
-	h.lock.Lock()
-	defer h.lock.Unlock()
+	go h.scaledObjectsMetricCache.Delete(key)
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:291]: okay, I will keep it for now, if we spot some problems in the future, we know where to look 😄

**Comment by v-shenoy** [comment]
@zroubalik I took a look at the PR, but it's kind of hard for me to understand the changes with 92 files changed (although I guess most of those are just merging the `IsActive` & `GetMetrics` methods for the scalers).

I am also a little bit confused by the PR title. Doesn't KEDA already query metrics on polling interval?

File: tests/internals/cache_metrics/cache_metrics_test.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, defaultMonitoredDeploymentReplicas, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	data.EnableUseCachedMetrics = true
+	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)
+
+	// Metric Value = 8, DesiredAverageMetricValue = 2
+	// should scale in to 8/2 = 4 replicas, irrespective of current replicas
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 4, 60, 1),
+		"replica count should be 4 after 1 minute")
```
Comment (JorTurFer) [inline tests/internals/cache_metrics/cache_metrics_test.go:164]: I'd give more time, something like 2 minutes or 3, this helper will finish once the target replica count is reached, and 1 minute having to scale to 4, could not be enough time

File: tests/internals/cache_metrics/cache_metrics_test.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	// initial replica count for a monitored deployment
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, defaultMonitoredDeploymentReplicas, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	data.EnableUseCachedMetrics = false
+	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)
+
+	// Metric Value = 8, DesiredAverageMetricValue = 2
+	// should scale in to 8/2 = 4 replicas, irrespective of current replicas
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 4, 60, 1),
```
Comment (JorTurFer) [inline tests/internals/cache_metrics/cache_metrics_test.go:199]: I'd give more time, something like 2 minutes or 3, this helper will finish once the target replica count is reached, and 1 minute having to scale to 4, could not be enough time

**Review by JorTurFer — APPROVED** [review]
LGTM! 2 suggestions inline, but not blockers

**Comment by champly** [comment]
> @champly could you please help me understand why is this test failing?
> 
> ```
> • Failure [5.076 seconds]
> ScaledObjectController
> /__w/keda/keda/controllers/keda/scaledobject_controller_test.go:54
>   scaleobject ready condition 'False/Unknow' to 'True' will requeue [It]
>   /__w/keda/keda/controllers/keda/scaledobject_controller_test.go:721
> 
>   Timed out after 5.000s.
>   Expected
>       <v1.ConditionStatus>: True
>   To satisfy at least one of these matchers: [%!s(*matchers.EqualMatcher=&{False}) %!s(*matchers.EqualMatcher=&{Unknown})]
> ```
> 
> you wrote the test and I am not sure, what could break the behavior, thanks!

On my local test it looks like it should work will.

**Comment by zroubalik** [comment]
> > @champly could you please help me understand why is this test failing?
> > ```
> > • Failure [5.076 seconds]
> > ScaledObjectController
> > /__w/keda/keda/controllers/keda/scaledobject_controller_test.go:54
> >   scaleobject ready condition 'False/Unknow' to 'True' will requeue [It]
> >   /__w/keda/keda/controllers/keda/scaledobject_controller_test.go:721
> > 
> >   Timed out after 5.000s.
> >   Expected
> >       <v1.ConditionStatus>: True
> >   To satisfy at least one of these matchers: [%!s(*matchers.EqualMatcher=&{False}) %!s(*matchers.EqualMatcher=&{Unknown})]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > you wrote the test and I am not sure, what could break the behavior, thanks!
> 
> On my local test it looks like it should work will.

@champly thanks a lot for the response, I have already solved this :)

**Comment by tomkerkhove** [comment]
Last run and I'll merge it @zroubalik

File: tests/internals/cache_metrics/cache_metrics_test.go (hunk 9 above)
Comment (tomkerkhove) [reply tests/internals/cache_metrics/cache_metrics_test.go:164]: @zroubalik Is this resolved?

File: tests/internals/cache_metrics/cache_metrics_test.go (hunk 10 above)
Comment (tomkerkhove) [reply tests/internals/cache_metrics/cache_metrics_test.go:199]: @zroubalik Is this resolved?

**Comment by zroubalik** [comment]
> Last run and I'll merge it @zroubalik

please let the merge on me, there might be some PRs merged before this one, so rebase might be needed.

**Review by JorTurFer — APPROVED** [review]
